Install windows binary releases

genotrance opened the pull request:
Much cleaner than before

dom96 commented on src/choosenim.nim:
```diff
@@ -30,13 +30,24 @@ proc parseVersion(versionStr: string): Version =
   removeDir(extractDir)
-  extract(path, extractDir)
+  if path.splitFile().ext == ".zip":
+    extract(path, extractDir.parentDir())
```
`extract` shouldn't have different behaviour based on the archive you're extracting, this if branch should be in the `extract`.

genotrance replied:
This is only because the zip file contents already has a `nim-$version` folder in it. Same will apply to xz files as well when we migrate to nimarchive. I can move it to `extract()` if preferred though.

dom96 commented on src/choosenim.nim:
```diff
@@ -30,13 +30,24 @@ proc parseVersion(versionStr: string): Version =
+      let pathEnv = getEnv("PATH")
+      putEnv("PATH", params.getMingwBin() & PathSep & pathEnv)
+      defer:
+        putEnv("PATH", pathEnv)
```
Please move this to the appropriate place in ``download`` (it should actually go into ``getGccArch`` I think?).

dom96 replied:
Even if it means code duplication in ``build``.

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+        display("Info:", "Binary build unavailable, building from source",
+                priority = HighPriority)
+
+    url = websiteUrl % $version
```
```suggestion
```

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+
+    var
+      url: string
+      outputPath: string
```
Move these definitions below the `when`

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+      url: string
+      outputPath: string
+    when defined(Windows):
+      url = winBinaryZipUrl % [$version, $arch]
```
```suggestion
      let winUrl = winBinaryZipUrl % [$version, $arch]
```

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+      outputPath: string
+    when defined(Windows):
+      url = winBinaryZipUrl % [$version, $arch]
+      if not needsDownload(params, url, outputPath): return outputPath
```
```suggestion
      if not needsDownload(params, winUrl, outputPath): return outputPath
```

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+      url = winBinaryZipUrl % [$version, $arch]
+      if not needsDownload(params, url, outputPath): return outputPath
+      try:
+        downloadFile(url, outputPath, params)
```
```suggestion
        downloadFile(winUrl, outputPath, params)
```

dom96 commented on src/choosenim/download.nim:
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
+    var
+      url: string
+      outputPath: string
+    when defined(Windows):
```
It seems like it would be relatively easy to check for the presence of `powershell` on the system so that we can continue to support Windows XP. Please do this by either:

* Using the ``osinfo`` package
* Running `powershell -v` (guessing the syntax here, just something that will always work on systems with powershell) and verifying it succeeds

If powershell isn't supported then we want to build from source.

genotrance commented:
I have addressed all your concerns - PTAL.

BTW, CI ran fine on my [instance](https://travis-ci.org/genotrance/choosenim/builds/479256174) of Travis but failed here for some reason.

genotrance commented:
Okay passed this time - looks like tester has some issue in detecting successful pass.

dom96 commented on src/choosenim/switcher.nim:
```diff
@@ -42,8 +42,12 @@ proc areProxiesInstalled(params: CliParams, proxies: openarray[string]): bool =
+  when defined(Windows):
+    return findExe("gcc") != ""
+  else:
+    # Fix issue #104
```
This statement is ambiguous, is it a TODO or a "This code fixes issue #104"?

I'm assuming the former, but you seem to already be sort of fixing it? Just fix it properly.

* For Windows: check for gcc only
* For macOS (or freebsd): check for clang only
* For Linux: check for gcc only

And maybe rename this function to `isDefaultCCInPath` since these are the default C compilers used by Nim on these platforms.

genotrance replied:
I was fixing it based on our IRC conversation. I've updated it.

dom96 commented on src/choosenim/builder.nim:
```diff
@@ -82,7 +82,9 @@ proc build*(extractDir: string, version: Version, params: CliParams) =
   let pathEnv = getEnv("PATH")
-  putEnv("PATH", params.getMingwBin() & PathSep & pathEnv)
+  when defined(windows):
+    if not isCCInPath(params) and dirExists(params.getMingwBin()):
```
Why is this necessary?

genotrance replied:
It is consistent with the other instance. And it is only needed on Windows so this makes it clear.

dom96 commented on src/choosenim/download.nim:
```diff
@@ -199,6 +200,15 @@ proc needsDownload(params: CliParams, downloadUrl: string,
     return false
 
 proc downloadImpl(version: Version, params: CliParams): string =
+  # Add MingW bin dir to PATH so getGccArch can find gcc.
```
without this ``getGccArch`` won't work, so this code should be inside ``getGccArch``

genotrance replied:
I put it in `downloadImpl()` to avoid having to propagate `params` all over the place and minimize changes. I've changed it again.

dom96 commented on src/choosenim.nim:
```diff
@@ -36,7 +36,12 @@ proc installVersion(version: Version, params: CliParams) =
   # Make sure no stale files from previous installation exist.
   removeDir(extractDir)
-  extract(path, extractDir)
+  if ("nim-$1_x" % $version) in path.splitFile().name:
```
Pretty sure this will break installation on other platforms. Also, my ``untar`` library already does this, so your new proc ``extractZip`` should do the same.

See how I do it in ``untar`` and do the same thing: https://github.com/dom96/untar/blob/master/src/untar.nim#L127 (also note that ``-j`` can be used for ``unzip``)

genotrance replied:
Note that `-j` will get rid of all directory structure, not just the base dir. It will extract all files into the destination as one big blob.

I'll do this in extractZip().
